[FLINK-703] Use complete element as join key

chiwanpark opened the pull request:
Hello. I open a pull request about FLINK-703. You can find more detail description in [JIRA](https://issues.apache.org/jira/browse/FLINK-703). This PR contains following changes.

* Add `BasicKeySelector` class to use complete element as key.
* Add `checkForAtomicType` method in `Keys` class to check condition.
* Modify `CoGroupOperator`, `JoinOperator`, `DataSet` in Java and Scala API
* Add some unit tests and integration tests to test this modification.

hsaputra commented on flink-java/src/main/java/org/apache/flink/api/java/functions/BasicKeySelector.java:
```diff
@@ -0,0 +1,26 @@
+
+package org.apache.flink.api.java.functions;
+
+public class BasicKeySelector<T> implements KeySelector<T, T> {
```
Please add JavaDoc explaining why the class is introduced and its relationships with others.

chiwanpark commented:
@hsaputra Thanks for advice! I renamed `BasicKeySelector` to `IdentityKeySelector` to prevent ambiguous naming. I added JavaDoc for `IdentityKeySelector`.

BTW, I found some duplicated classes with `IdentityKeySelector` ([1](https://github.com/apache/flink/blob/master/flink-optimizer/src/test/java/org/apache/flink/optimizer/testfunctions/IdentityKeyExtractor.java), [2](https://github.com/apache/flink/blob/master/flink-tests/src/test/java/org/apache/flink/test/iterative/StaticlyNestedIterationsITCase.java#L58), [3](https://github.com/apache/flink/blob/master/flink-tests/src/test/java/org/apache/flink/test/misc/CustomPartitioningITCase.java#L67)). Should I substitute the `IdentityKeySelector` for them? It will decrease the amount of duplicated code.

mxm commented:
Looks good @chiwanpark. Could you add some documentation besides the java doc? For example, here: http://ci.apache.org/projects/flink/flink-docs-master/programming_guide.html#define-keys-using-key-selector-functions

fhueske commented:
Thanks @chiwanpark for the PR!

Using an IdentityKeySelector is not the best solution in this case. A KeySelector<X,Y> transparently converts a `DataSet<X>` into a `DataSet<Tuple2<Y,X>` and uses the first tuple field as key. Since the KeySelector used here is an IdentityKeySelector we end up with a `DataSet<X,X>` which unnecessarily doubles the amount of data. 

I will look at this PR later in detail and give some feedback how it could be improved. Thanks!

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -710,6 +711,9 @@ public long count() throws Exception {
 	 * @see DataSet
 	 */
 	public UnsortedGrouping<T> groupBy(String... fields) {
+		if (Keys.checkForAtomicType(getType(), fields)) {
```
We need to check a few things here.
- If atomic type, we only allow 
  1. a single expression (`fields.length`)
  1. the expression must be a wildcard expression `"*"`
  1. the atomic type must be a key type (`TypeInformation.isKeyType() == true`)
- else do as before

fhueske replied:
Sorry, just saw that you are doing exactly that in `checkForAtomicType()`, very good! :-)

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/functions/IdentityKeySelector.java:
```diff
@@ -0,0 +1,42 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.functions;
```
As said before, we should do this without an IdentityKeySelector.

fhueske replied:
A better solution than using a KeySelector would be to extend the Keys.ExpressionKey class to support AtomicTypes. The code from `checkForAtomicType()` could be moved to the constructor of `Keys.ExpressionKey`. The `Keys.ExpressionKey.keyFields` member should be a list with a single `FlatFieldDescriptor` that points to flat position `0` and have the type of the AtomicType.

Generalizing the ExpressionKey to AtomicType should not have side effects, but you should run a build with all tests to verify that.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/Keys.java:
```diff
@@ -410,6 +410,11 @@ public String toString() {
 	}
+
+	public static <T> boolean checkForAtomicType(TypeInformation<T> typeInfo, String... fields) {
+		return typeInfo.isKeyType() && typeInfo instanceof AtomicType && fields.length == 1 &&
```
The checks in this method are correct.
I would structure it a bit different though, to be able to give meaningful error messages.
First check if the type is an AtomicType, if no return false.
Otherwise, do the other checks and throw an `InvalidProgramException` with a meaningful error message if any of them is not true.

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/operator/GroupingTest.java:
```diff
@@ -551,6 +551,14 @@ public Long getKey(Tuple4<Integer, Long, CustomType, Long[]> value) throws Excep
+
+		dataSet.groupBy("*");
+	}
+
```
Please also add tests that check for correct behavior if users define invalid programs (more than one expression, no wildcard, no key type).

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/operator/JoinOperatorTest.java:
```diff
@@ -585,6 +585,24 @@ public Long getKey(CustomType value) {
+		DataSet<Integer> ds2 = env.fromElements(0, 0, 0);
+
+		ds1.join(ds2).where(0).equalTo("*");
+	}
```
Please add tests that check for correct handling of incorrect programs.

fhueske commented on flink-tests/src/test/scala/org/apache/flink/api/scala/operators/CoGroupOperatorTest.scala:
```diff
@@ -268,6 +268,24 @@ class CoGroupOperatorTest {
+    val ds2 = env.fromCollection(emptyTupleData)
+
+    ds1.coGroup(ds2).where("*").equalTo(0)
+  }
```
Please add tests for correct behavior in case of incorrect programs.

fhueske commented on flink-tests/src/test/scala/org/apache/flink/api/scala/operators/GroupingTest.scala:
```diff
@@ -224,5 +224,13 @@ class GroupingTest {
+    val ds = env.fromElements(0, 1, 2)
+
+    ds.groupBy("*")
+  }
```
Please add tests for correct behavior in case of incorrect programs.

fhueske commented on flink-tests/src/test/scala/org/apache/flink/api/scala/operators/JoinOperatorTest.scala:
```diff
@@ -272,5 +268,14 @@ class JoinOperatorTest {
+    val ds2 = env.fromCollection(emptyLongData)
+
+    ds1.join(ds2).where(1).equalTo("*")
+  }
```
Please add tests for correct behavior in case of incorrect programs.

fhueske commented:
Very good first start! 
The code looks correct and there are many tests, nice :-)

However, as I said before I would not use KeySelectors because they basically double the data. Generalizing `ExpressionKeys` for AtomicTypes seems to be the better approach.
Also some tests for checking correct behavior in case of invalid programs would be good to have.

chiwanpark commented:
@fhueske Thank you for review! I will reimplement without KeySelector and update the PR.
@mxm Okay, I will add some documentation after reimplement this. Thank you.

chiwanpark commented:
Unfortunately, Because I'm undergraduate student, I have to put in my time on preparing mid-exam. :(
I will update the PR as soon as possible. Maybe the PR will be updated in next week.

fhueske commented:
No worries, that's absolutely fine :-)
Good luck with your exams!

chiwanpark commented:
Hi. I updated this PR. The changes are following.

* Re-implement this feature with generalizing `ExpressionKeys`.
* Modify `CoGroupOperatorBase`, `GroupCombineOperatorBase` and `GroupReduceOperatorBase` to allow use of AtomicType as Key.
* Add unit tests to test invalid usage.

Because a mention for wildcard expression with atomic type exists already in documentation. I didn't modify documentation.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/GroupReduceITCase.java:
```diff
@@ -1063,6 +1065,33 @@ public void reduce(Iterable<Tuple5<Integer, Long, Integer, String, Long>> values
+		DataSet<Integer> reduceDs = ds.groupBy("*").reduceGroup(new GroupReduceFunction<Integer, Integer>() {
+			@Override
+			public void reduce(Iterable<Integer> values, Collector<Integer> out) throws Exception {
+				Set<Integer> set = new HashSet<Integer>();
```
Since all elements of a group are identical when you do `groupBy("*")` this could be implemented a bit easier by just returning the first element of a group (`out.collect(values.iterator().next())`).

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/GroupReduceITCase.java:
```diff
@@ -1063,6 +1065,33 @@ public void reduce(Iterable<Tuple5<Integer, Long, Integer, String, Long>> values
+			}
+		});
+
+		env.setParallelism(1);
```
Why do you set the parallelism to 1?

chiwanpark replied:
Oh, I just copied code from other test case. I will fix it :)

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/JoinITCase.java:
```diff
@@ -663,6 +663,39 @@ public void testNonPojoToVerifyNestedTupleElementSelectionWithFirstKeyFieldGreat
+		DataSet<Tuple2<Tuple3<Integer, Long, String>, Integer>> joinDs = ds1.join(ds2).where(0).equalTo("*");
+
+		joinDs.writeAsCsv(resultPath);
+		env.setParallelism(1);
```
Why parallelism == 1?

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/JoinITCase.java:
```diff
@@ -663,6 +663,39 @@ public void testNonPojoToVerifyNestedTupleElementSelectionWithFirstKeyFieldGreat
+		DataSet<Tuple2<Integer, Tuple3<Integer, Long, String>>> joinDs = ds1.join(ds2).where("*").equalTo(0);
+
+		joinDs.writeAsCsv(resultPath);
+		env.setParallelism(1);
```
Why parallelism == 1?

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/Keys.java:
```diff
@@ -274,24 +274,31 @@ private static int countNestedElementsBefore(CompositeType<?> compositeType, int
-				List<FlatFieldDescriptor> keys = cType.getFlatFields(expressions[i]); // use separate list to do a size check
-				if(keys.size() == 0) {
-					throw new IllegalArgumentException("Unable to extract key from expression '"+expressions[i]+"' on key "+cType);
+			if (type instanceof AtomicType) {
```
can you add a `null` check for `expressionsIn`?

fhueske commented:
Thanks for the update. Looks really good! 
I have just a few inline comments. Once these are resolved, I would try the code and merge it if everything works.

chiwanpark commented:
I updated this PR. Thanks for advice. :)

* Remove `setParallelism(1)` in test code.
* Simplify `testGroupReduceWithAtomicValue` in `GroupReduceITCase`.
* Add a `null` check for `expressionsIn` in constructor of `ExpressionKeys`.

fhueske commented:
Thanks for the prompt update and the good work!
Will try the PR and merge it if everything works fine :-)
